Monads.NET/Maybe.cs

```diff
@@ -162,6 +197,37 @@ public static TInput IfDo<TInput>(this TInput o, Func<TInput, bool> evaluator, A
         }
 
         /// <summary>
+        /// Allows for a null safe action on a object.
+        /// </summary>
+        /// <typeparam name="TInput">Instance of object to be acted on</typeparam>
+        /// <param name="o">object to be extended</param>
+        /// <param name="action">function that will act on the object</param>
+        /// <returns>object that was acted on</returns>
+        public static Task DoAsync<TInput>(this TInput o, Func<TInput, Task> action)
+            where TInput : class
+        {
+            return (o == null) ? TaskEx.FromResult<object>(null) : action(o);
+        }
+
+        /// <summary>
+        /// Allows for a null safe action on a IEnumerable of objects.
+        /// </summary>
+        /// <typeparam name="TInput">Instance of object to be acted on</typeparam>
+        /// <param name="o">collection to be extended</param>
+        /// <param name="action">function that will act on each object</param>
+        /// <returns>object that was acted on</returns>
+        public static async Task<IEnumerable<TInput>> DoAsync<TInput>(this IEnumerable<TInput> o, Func<TInput, Task> action) where TInput : class
+        {
+            if (o == null) return null;
+
+            foreach (var c in o)
+            {
+                await action(c);
+            }
+            return o;
```

Comment:
Is there a reason these are executed sequentially instead of in parallel? Same question as the use of `TaskEx.WhenAll` really. I suggest the method names be changed to be descriptive of this behavior, e.g. `DoAsyncSequential` and `DoAsyncConcurrent`